libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectEx.kt

```diff
@@ -2,68 +2,66 @@
-import com.mapbox.navigation.core.trip.roadobject.reststop.RestStopFromIntersection
-import com.mapbox.navigation.core.trip.roadobject.reststop.RestStopFromRoadObject
+import com.mapbox.navigation.core.navigator.toUpcomingRoadObjects
 import com.mapbox.navigation.utils.internal.ifNonNull
+import com.mapbox.navigator.RouteInterface
 
 /**
- * Use the API to retrieve a list of all upcoming rest stops using [UpcomingRoadObject].
- * This is recommended to be used with
- * - [RouteProgress] during Active Navigation.
+ * Use the API to retrieve the first upcoming [UpcomingRoadObject].
+ * - given that route has no [UpcomingRoadObject] of type [clazz], returns null
+ * - given that route has 1 [UpcomingRoadObject] of type [clazz], returns it considering
+ *   [UpcomingRoadObject.distanceToStart] is positive
+ * - given that route has 1 [UpcomingRoadObject] of type [clazz], returns null considering
+ *   [UpcomingRoadObject.distanceToStart] is either 0, negative or null
+ * - given that route has multiple upcoming road objects of type [clazz], returns the first
+ *   [UpcomingRoadObject] for which [UpcomingRoadObject.distanceToStart] is positive
+ * - given that route has multiple upcoming road objects of type [clazz], returns null if none has
+ *   a positive [UpcomingRoadObject.distanceToStart]
+ *
+ * @param type determines [RoadObjectType] to filter. Any other value will return null.
  */
-fun List<UpcomingRoadObject>.getAllRestStops(): List<RestStopFromRoadObject> {
-    return RestStopFactory.createWithUpcomingRoadObjects(this)
-}
-
-/**
- * Use the API to retrieve the first upcoming rest stop using [UpcomingRoadObject]. The API will
- * always find the first upcoming rest stop from the list of upcoming rest stops if available.
- * - returns null if route has no rest stop: In this case the API will return null
- * - returns [RestStopFromRoadObject] if route has 1 rest stop and hasn't been passed/reached
- * - returns null if route has 1 rest stop and it has already been passed/reached
- * - returns the first [RestStopFromRoadObject] that hasn't been passed/reached if route has multiple
- *   rest stops and you have passed a few of them
- * - returns null if route has multiple rest stops and all have been passed
- * This is recommended to be used with
- * - [RouteProgress] during Active Navigation.
- */
-fun List<UpcomingRoadObject>.getFirstUpcomingRestStop(): RestStopFromRoadObject? {
-    val restStops = this.getAllRestStops()
-    if (restStops.isNotEmpty()) {
-        val distanceToStart = restStops.first().distanceToStart
+fun RouteProgress.getFirstUpcomingRoadObject(@RoadObjectType.Type type: Int): UpcomingRoadObject? {
+    return ifNonNull(this.upcomingRoadObjects.firstOrNull { it.roadObject.objectType == type }) {
+        val distanceToStart = it.distanceToStart
         if (distanceToStart != null && distanceToStart > 0) {
-            return restStops.first()
+            it
+        } else {
+            null
         }
```

Comment:
I changed this extension to 

```
return this.upcomingRoadObjects.firstOrNull {
        if (it.roadObject.objectType == type) {
            val distanceToStart = it.distanceToStart
            distanceToStart != null && distanceToStart > 0
        } else {
            false
        }
    }
```
I understand your point, but similar to you 
> It might still return tunnels with negative distanceToStart while we're in them (not 100% sure), but nothing more than that.
I am unsure of this. Hence, I feel with the above changed logic, there is no harm in having an extension. This rightfully only returns that `UpcomingRoadObject` which you are interested in and has not already passed. So, even if in some cases the `List<UpcomingRoadObject>` might contain `UpcomingRoadObject` with -ve/0/null `distanceToStart` the extension won't return it.

Wdyt about leaving this? Anyways, it's a simple extension. Customers can use their own logic if they have a different use case.

cc @LukasPaczos

Reply:
The documentation for `UpcomingRoadObject` explains:
```
 * @param distanceToStart remaining distance to the start of the object.
 * This value will be negative after passing the start of the object and until we cross the finish
 * point of the [RoadObject]s geometry for objects that are on the actively navigated route,
 * but it will be zero for [EHorizon] objects. It will be null if couldn't be determined.
```

I don't see a need for adding an additional extension just to check the distance. Consumers should be in a position to do the simple check when needed, depending on their needs.

Reply:
🤷🏻 Ok removing the extension.